File: android-gif-drawable/src/main/java/pl/droidsonroids/gif/GifInfoHandle.java
package pl.droidsonroids.gif;

import android.content.ContentResolver;
import android.content.res.AssetFileDescriptor;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;

import androidx.annotation.FloatRange;
import androidx.annotation.IntRange;
import androidx.annotation.RequiresApi;

import android.system.ErrnoException;
import android.system.Os;
import android.view.Surface;

import java.io.FileDescriptor;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;

/**
 * Native library wrapper
 */
final class GifInfoHandle {
	static {
		LibraryLoader.loadLibrary();
	}

	/**
	 * Pointer to native structure. Access must be synchronized, heap corruption may occur otherwise
	 * when {@link #recycle()} is called during another operation.
	 */
	private volatile long gifInfoPtr;

	GifInfoHandle() {
	}

	GifInfoHandle(FileDescriptor fileDescriptor) throws GifIOException {
		gifInfoPtr = openFileDescriptor(fileDescriptor, 0, true);
	}

	GifInfoHandle(byte[] bytes) throws GifIOException {
		gifInfoPtr = openByteArray(bytes);
	}

	GifInfoHandle(ByteBuffer buffer) throws GifIOException {
		gifInfoPtr = openDirectByteBuffer(buffer);
	}

	GifInfoHandle(String filePath) throws GifIOException {
		gifInfoPtr = openFile(filePath);
	}

	GifInfoHandle(InputStream stream) throws GifIOException {
		if (!stream.markSupported()) {
			throw new IllegalArgumentException("InputStream does not support marking");
		}
		gifInfoPtr = openStream(stream);
	}

	GifInfoHandle(AssetFileDescriptor afd) throws IOException {
		try {
			gifInfoPtr = openFileDescriptor(afd.getFileDescriptor(), afd.getStartOffset(), false);
		} finally {
			try {
				afd.close();
			} catch (IOException ignored) {
				//no-op
			}
		}
	}

	private static long openFileDescriptor(FileDescriptor fileDescriptor, long offset, boolean closeOriginalDescriptor) throws GifIOException {
		final int nativeFileDescriptor;
		if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O_MR1) {
			try {
				nativeFileDescriptor = getNativeFileDescriptor(fileDescriptor, closeOriginalDescriptor);
			} catch (Exception e) { //cannot catch ErrnoException due to VerifyError on API <= 19
				throw new GifIOException(GifError.OPEN_FAILED.errorCode, e.getMessage());
			}
		} else {
			nativeFileDescriptor = extractNativeFileDescriptor(fileDescriptor, closeOriginalDescriptor);
		}
		return openNativeFileDescriptor(nativeFileDescriptor, offset);
	}

	@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
	private static int getNativeFileDescriptor(FileDescriptor fileDescriptor, boolean closeOriginalDescriptor) throws GifIOException, ErrnoException {
		try {
			final int nativeFileDescriptor = createTempNativeFileDescriptor();
			Os.dup2(fileDescriptor, nativeFileDescriptor);
			return nativeFileDescriptor;
		} finally {
			if (closeOriginalDescriptor) {
				Os.close(fileDescriptor);
			}
		}
	}

	static GifInfoHandle openUri(ContentResolver resolver, Uri uri) throws IOException {
		if (ContentResolver.SCHEME_FILE.equals(uri.getScheme())) { //workaround for #128
			return new GifInfoHandle(uri.getPath());
		}
		final AssetFileDescriptor assetFileDescriptor = resolver.openAssetFileDescriptor(uri, "r");
		if (assetFileDescriptor == null) {
			throw new IOException("Could not open AssetFileDescriptor for " + uri);
		}
		return new GifInfoHandle(assetFileDescriptor);
	}

	static native long openNativeFileDescriptor(int fd, long offset) throws GifIOException;

	static native int extractNativeFileDescriptor(FileDescriptor fileDescriptor, boolean closeOriginalDescriptor) throws GifIOException;

	static native int createTempNativeFileDescriptor() throws GifIOException;

	static native long openByteArray(byte[] bytes) throws GifIOException;

	static native long openDirectByteBuffer(ByteBuffer buffer) throws GifIOException;

	static native long openStream(InputStream stream) throws GifIOException;

	static native long openFile(String filePath) throws GifIOException;

	private static native long renderFrame(long gifFileInPtr, Bitmap frameBuffer);

	private static native void bindSurface(long gifInfoPtr, Surface surface, long[] savedState);

	private static native void free(long gifFileInPtr);

	private static native boolean reset(long gifFileInPtr);

	private static native void setSpeedFactor(long gifFileInPtr, float factor);

	private static native String getComment(long gifFileInPtr);

	private static native int getLoopCount(long gifFileInPtr);

	private static native void setLoopCount(long gifFileInPtr, char loopCount);

	private static native long getSourceLength(long gifFileInPtr);

	private static native int getDuration(long gifFileInPtr);

	private static native int getCurrentPosition(long gifFileInPtr);

	private static native void seekToTime(long gifFileInPtr, int position, Bitmap buffer);

	private static native void seekToFrame(long gifFileInPtr, int frameNr, Bitmap buffer);

	private static native void saveRemainder(long gifFileInPtr);

	private static native long restoreRemainder(long gifFileInPtr);

	private static native long getAllocationByteCount(long gifFileInPtr);

	private static native long getMetadataByteCount(long gifFileInPtr);

	private static native int getNativeErrorCode(long gifFileInPtr);

	private static native int getCurrentFrameIndex(long gifFileInPtr);

	private static native int getCurrentLoop(long gifFileInPtr);

	private static native void postUnbindSurface(long gifFileInPtr);

	private static native boolean isAnimationCompleted(long gifInfoPtr);

	private static native long[] getSavedState(long gifInfoPtr);

	private static native int restoreSavedState(long gifInfoPtr, long[] savedState, Bitmap mBuffer);

	private static native int getFrameDuration(long gifInfoPtr, int index);

	private static native void setOptions(long gifInfoPtr, char sampleSize, boolean isOpaque);

	private static native int getWidth(long gifFileInPtr);

	private static native int getHeight(long gifFileInPtr);

	private static native int getNumberOfFrames(long gifInfoPtr);

	private static native boolean isOpaque(long gifInfoPtr);

	private static native void startDecoderThread(long gifInfoPtr);

	private static native void stopDecoderThread(long gifInfoPtr);

	private static native void glTexImage2D(long gifInfoPtr, int target, int level);

	private static native void glTexSubImage2D(long gifInfoPtr, int target, int level);

	private static native void seekToFrameGL(long gifInfoPtr, int index);

	private static native void initTexImageDescriptor(long gifInfoPtr);

	synchronized long renderFrame(Bitmap frameBuffer) {
		return renderFrame(gifInfoPtr, frameBuffer);
	}

	void bindSurface(Surface surface, long[] savedState) {
		bindSurface(gifInfoPtr, surface, savedState);
	}

	synchronized void recycle() {
		free(gifInfoPtr);
		gifInfoPtr = 0L;
	}

	synchronized long restoreRemainder() {
		return restoreRemainder(gifInfoPtr);
	}

	synchronized boolean reset() {
		return reset(gifInfoPtr);
	}

	synchronized void saveRemainder() {
		saveRemainder(gifInfoPtr);
	}

	synchronized String getComment() {
		return getComment(gifInfoPtr);
	}

	synchronized int getLoopCount() {
		return getLoopCount(gifInfoPtr);
	}

	void setLoopCount(@IntRange(from = 0, to = Character.MAX_VALUE) final int loopCount) {
		if (loopCount < 0 || loopCount > Character.MAX_VALUE) {
			throw new IllegalArgumentException("Loop count of range <0, 65535>");
		}
		synchronized (this) {
			setLoopCount(gifInfoPtr, (char) loopCount);
		}
	}

	synchronized long getSourceLength() {
		return getSourceLength(gifInfoPtr);
	}

	synchronized int getNativeErrorCode() {
		return getNativeErrorCode(gifInfoPtr);
	}

	void setSpeedFactor(@FloatRange(from = 0, fromInclusive = false) float factor) {
		if (factor <= 0f || Float.isNaN(factor)) {
			throw new IllegalArgumentException("Speed factor is not positive");
		}
		if (factor < 1f / Integer.MAX_VALUE) {
			factor = 1f / Integer.MAX_VALUE;
		}
		synchronized (this) {
			setSpeedFactor(gifInfoPtr, factor);
		}
	}

	synchronized int getDuration() {
		return getDuration(gifInfoPtr);
	}

	synchronized int getCurrentPosition() {
		return getCurrentPosition(gifInfoPtr);
	}

	synchronized int getCurrentFrameIndex() {
		return getCurrentFrameIndex(gifInfoPtr);
	}

	synchronized int getCurrentLoop() {
		return getCurrentLoop(gifInfoPtr);
	}

	synchronized void seekToTime(@IntRange(from = 0, to = Integer.MAX_VALUE) final int position, final Bitmap buffer) {
		seekToTime(gifInfoPtr, position, buffer);
	}

	synchronized void seekToFrame(@IntRange(from = 0, to = Integer.MAX_VALUE) final int frameIndex,
			final Bitmap buffer) {
		seekToFrame(gifInfoPtr, frameIndex, buffer);
	}

	synchronized long getAllocationByteCount() {
		return getAllocationByteCount(gifInfoPtr);
	}

	synchronized long getMetadataByteCount() {
		return getMetadataByteCount(gifInfoPtr);
	}

	synchronized boolean isRecycled() {
		return gifInfoPtr == 0L;
	}

	@Override
	protected void finalize() throws Throwable {
		try {
			recycle();
		} finally {
			super.finalize();
		}
	}

	synchronized void postUnbindSurface() {
		postUnbindSurface(gifInfoPtr);
	}

	synchronized boolean isAnimationCompleted() {
		return isAnimationCompleted(gifInfoPtr);
	}

	synchronized long[] getSavedState() {
		return getSavedState(gifInfoPtr);
	}

	synchronized int restoreSavedState(long[] savedState, Bitmap mBuffer) {
		return restoreSavedState(gifInfoPtr, savedState, mBuffer);
	}

	synchronized int getFrameDuration(@IntRange(from = 0) final int index) {
		throwIfFrameIndexOutOfBounds(index);
		return getFrameDuration(gifInfoPtr, index);
	}

	void setOptions(char sampleSize, boolean isOpaque) {
		setOptions(gifInfoPtr, sampleSize, isOpaque);
	}

	synchronized int getWidth() {
		return getWidth(gifInfoPtr);
	}

	synchronized int getHeight() {
		return getHeight(gifInfoPtr);
	}

	synchronized int getNumberOfFrames() {
		return getNumberOfFrames(gifInfoPtr);
	}

	synchronized boolean isOpaque() {
		return isOpaque(gifInfoPtr);
	}

	void glTexImage2D(int target, int level) {
		glTexImage2D(gifInfoPtr, target, level);
	}

	void glTexSubImage2D(int target, int level) {
		glTexSubImage2D(gifInfoPtr, target, level);
	}

	void startDecoderThread() {
		startDecoderThread(gifInfoPtr);
	}

	void stopDecoderThread() {
		stopDecoderThread(gifInfoPtr);
	}

	void initTexImageDescriptor() {
		initTexImageDescriptor(gifInfoPtr);
	}

	void seekToFrameGL(@IntRange(from = 0) final int index) {
		throwIfFrameIndexOutOfBounds(index);
		seekToFrameGL(gifInfoPtr, index);
	}

	private void throwIfFrameIndexOutOfBounds(@IntRange(from = 0) final int index) {
		final int numberOfFrames = getNumberOfFrames(gifInfoPtr);
		if (index < 0 || index >= numberOfFrames) {
			throw new IndexOutOfBoundsException("Frame index is not in range <0;" + numberOfFrames + '>');
		}
	}
}

File: android-gif-drawable/src/main/java/pl/droidsonroids/gif/GifInfoHandle.kt
package pl.droidsonroids.gif

import android.content.ContentResolver
import android.content.res.AssetFileDescriptor
import android.graphics.Bitmap
import android.net.Uri
import android.os.Build
import android.system.ErrnoException
import android.system.Os
import android.view.Surface
import androidx.annotation.FloatRange
import androidx.annotation.IntRange
import androidx.annotation.RequiresApi
import pl.droidsonroids.gif.LibraryLoader.loadLibrary
import java.io.FileDescriptor
import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer
import kotlin.Boolean
import kotlin.ByteArray
import kotlin.Char
import kotlin.Exception
import kotlin.IndexOutOfBoundsException
import kotlin.Int
import kotlin.Long
import kotlin.LongArray
import kotlin.String
import kotlin.Throwable
import kotlin.Throws
import kotlin.code
import kotlin.require
import kotlin.synchronized

/**
 * Native library wrapper
 */
class GifInfoHandle {
    /**
     * Pointer to native structure. Access must be synchronized, heap corruption may occur otherwise
     * when [.recycle] is called during another operation.
     */
    @Volatile
    private var gifInfoPtr: Long = 0

    constructor()
    constructor(fileDescriptor: FileDescriptor) {
        gifInfoPtr = openFileDescriptor(fileDescriptor, 0, true)
    }

    constructor(bytes: ByteArray?) {
        gifInfoPtr = openByteArray(bytes)
    }

    constructor(buffer: ByteBuffer?) {
        gifInfoPtr = openDirectByteBuffer(buffer)
    }

    constructor(filePath: String?) {
        gifInfoPtr = openFile(filePath)
    }

    constructor(stream: InputStream) {
        require(stream.markSupported()) { "InputStream does not support marking" }
        gifInfoPtr = openStream(stream)
    }

    constructor(afd: AssetFileDescriptor) {
        gifInfoPtr = try {
            openFileDescriptor(afd.fileDescriptor, afd.startOffset, false)
        } finally {
            try {
                afd.close()
            } catch (ignored: IOException) {
                //no-op
            }
        }
    }

    @Synchronized
    fun renderFrame(frameBuffer: Bitmap?): Long {
        return renderFrame(gifInfoPtr, frameBuffer)
    }

    fun bindSurface(surface: Surface, savedState: LongArray?) {
        bindSurface(gifInfoPtr, surface, savedState)
    }

    @Synchronized
    fun recycle() {
        free(gifInfoPtr)
        gifInfoPtr = 0L
    }

    @Synchronized
    fun restoreRemainder(): Long {
        return restoreRemainder(gifInfoPtr)
    }

    @Synchronized
    fun reset(): Boolean {
        return reset(gifInfoPtr)
    }

    @Synchronized
    fun saveRemainder() {
        saveRemainder(gifInfoPtr)
    }

    @get:Synchronized
    val comment: String
        get() = getComment(gifInfoPtr)

    @get:Synchronized
    var loopCount: Int
        get() = getLoopCount(gifInfoPtr)
        set(loopCount) {
            require(!(loopCount < 0 || loopCount > Character.MAX_VALUE.code)) { "Loop count of range <0, 65535>" }
            synchronized(this) { setLoopCount(gifInfoPtr, loopCount.toChar()) }
        }

    @get:Synchronized
    val inputSourceByteCount: Long
        get() = getSourceLength(gifInfoPtr)
/**
 * Returns length of the input source obtained at the opening time or -1 if
 * length cannot be determined. Returned value does not change during runtime.
 * If GifDrawable is constructed from [InputStream] -1 is always returned.
 * In case of byte array and [ByteBuffer] length is always known.
 * In other cases length -1 can be returned if length cannot be determined.
 *
 * @return number of bytes backed by input source or -1 if it is unknown
 */

    @get:Synchronized
    val nativeErrorCode: Int
        get() = getNativeErrorCode(gifInfoPtr)

    fun setSpeedFactor(@FloatRange(from = 0.0, fromInclusive = false) factor: Float) {
        var floatFactor = factor
        require(!(floatFactor <= 0f || floatFactor.isNaN())) { "Speed factor is not positive" }
        if (floatFactor < 1f / Int.MAX_VALUE) {
            floatFactor = 1f / Int.MAX_VALUE
        }
        synchronized(this) { setSpeedFactor(gifInfoPtr, floatFactor) }
    }

    @get:Synchronized
    val duration: Int
        get() = getDuration(gifInfoPtr)

    @get:Synchronized
    val currentPosition: Int
        get() = getCurrentPosition(gifInfoPtr)

    @get:Synchronized
    val currentFrameIndex: Int
        get() = getCurrentFrameIndex(gifInfoPtr)

    @get:Synchronized
    val currentLoop: Int
        get() = getCurrentLoop(gifInfoPtr)

    @Synchronized
    fun seekToTime(@IntRange(from = 0, to = Int.MAX_VALUE.toLong()) position: Int, buffer: Bitmap) {
        seekToTime(gifInfoPtr, position, buffer)
    }

    @Synchronized
    fun seekToFrame(
        @IntRange(from = 0, to = Int.MAX_VALUE.toLong()) frameIndex: Int,
        buffer: Bitmap
    ) {
        seekToFrame(gifInfoPtr, frameIndex, buffer)
    }

    @get:Synchronized
    val allocationByteCount: Long
        get() = getAllocationByteCount(gifInfoPtr)

    @get:Synchronized
    val metadataAllocationByteCount: Long
        get() = getMetadataByteCount(gifInfoPtr)

    @get:Synchronized
    val isRecycled: Boolean
        get() = gifInfoPtr == 0L

    @Throws(Throwable::class)
    protected fun finalize() {
        recycle()
    }

    @Synchronized
    fun postUnbindSurface() {
        postUnbindSurface(gifInfoPtr)
    }

    @get:Synchronized
    val isAnimationCompleted: Boolean
        get() = isAnimationCompleted(gifInfoPtr)

    @get:Synchronized
    val savedState: LongArray
        get() = getSavedState(gifInfoPtr)

    @Synchronized
    fun restoreSavedState(savedState: LongArray, mBuffer: Bitmap?): Int {
        return restoreSavedState(gifInfoPtr, savedState, mBuffer)
    }

    @Synchronized
    fun getFrameDuration(@IntRange(from = 0) index: Int): Int {
        throwIfFrameIndexOutOfBounds(index)
        return getFrameDuration(gifInfoPtr, index)
    }

    fun setOptions(sampleSize: Char, isOpaque: Boolean) {
        setOptions(gifInfoPtr, sampleSize, isOpaque)
    }

    @get:Synchronized
    val width: Int
        get() = getWidth(gifInfoPtr)

    @get:Synchronized
    val height: Int
        get() = getHeight(gifInfoPtr)

    @get:Synchronized
    val numberOfFrames: Int
        get() = getNumberOfFrames(gifInfoPtr)

    @get:Synchronized
    val isOpaque: Boolean
        get() = isOpaque(gifInfoPtr)

    fun glTexImage2D(target: Int, level: Int) {
        glTexImage2D(gifInfoPtr, target, level)
    }

    fun glTexSubImage2D(target: Int, level: Int) {
        glTexSubImage2D(gifInfoPtr, target, level)
    }

    fun startDecoderThread() {
        startDecoderThread(gifInfoPtr)
    }

    fun stopDecoderThread() {
        stopDecoderThread(gifInfoPtr)
    }

    fun initTexImageDescriptor() {
        initTexImageDescriptor(gifInfoPtr)
    }

    fun seekToFrameGL(@IntRange(from = 0) index: Int) {
        throwIfFrameIndexOutOfBounds(index)
        seekToFrameGL(gifInfoPtr, index)
    }

    private fun throwIfFrameIndexOutOfBounds(@IntRange(from = 0) index: Int) {
        val numberOfFrames = getNumberOfFrames(gifInfoPtr)
        if (index < 0 || index >= numberOfFrames) {
            throw IndexOutOfBoundsException("Frame index is not in range <0;$numberOfFrames>")
        }
    }

    companion object {
        init {
            loadLibrary()
        }

        @Throws(GifIOException::class)
        private fun openFileDescriptor(
            fileDescriptor: FileDescriptor,
            offset: Long,
            closeOriginalDescriptor: Boolean
        ): Long {
            val nativeFileDescriptor = if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O_MR1) {
                try {
                    getNativeFileDescriptor(fileDescriptor, closeOriginalDescriptor)
                } catch (e: Exception) { //cannot catch ErrnoException due to VerifyError on API <= 19
                    throw GifIOException(GifError.OPEN_FAILED.errorCode, e.message)
                }
            } else {
                extractNativeFileDescriptor(fileDescriptor, closeOriginalDescriptor)
            }
            return openNativeFileDescriptor(nativeFileDescriptor, offset)
        }

        @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
        @Throws(
            GifIOException::class, ErrnoException::class
        )
        private fun getNativeFileDescriptor(
            fileDescriptor: FileDescriptor,
            closeOriginalDescriptor: Boolean
        ): Int {
            return try {
                val nativeFileDescriptor = createTempNativeFileDescriptor()
                Os.dup2(fileDescriptor, nativeFileDescriptor)
                nativeFileDescriptor
            } finally {
                if (closeOriginalDescriptor) {
                    Os.close(fileDescriptor)
                }
            }
        }

        @Throws(IOException::class)
        fun openUri(resolver: ContentResolver, uri: Uri): GifInfoHandle {
            if (ContentResolver.SCHEME_FILE == uri.scheme) { //workaround for #128
                return GifInfoHandle(uri.path)
            }
            val assetFileDescriptor = resolver.openAssetFileDescriptor(uri, "r")
                ?: throw IOException("Could not open AssetFileDescriptor for $uri")
            return GifInfoHandle(assetFileDescriptor)
        }

        @Throws(GifIOException::class)
        external fun openNativeFileDescriptor(fd: Int, offset: Long): Long
        @Throws(GifIOException::class)
        external fun extractNativeFileDescriptor(
            fileDescriptor: FileDescriptor?,
            closeOriginalDescriptor: Boolean
        ): Int

        @Throws(GifIOException::class)
        external fun createTempNativeFileDescriptor(): Int
        @Throws(GifIOException::class)
        external fun openByteArray(bytes: ByteArray?): Long
        @Throws(GifIOException::class)
        external fun openDirectByteBuffer(buffer: ByteBuffer?): Long
        @Throws(GifIOException::class)
        external fun openStream(stream: InputStream?): Long
        @Throws(GifIOException::class)
        external fun openFile(filePath: String?): Long
        private external fun renderFrame(gifFileInPtr: Long, frameBuffer: Bitmap?): Long
        private external fun bindSurface(gifInfoPtr: Long, surface: Surface, savedState: LongArray?)
        private external fun free(gifFileInPtr: Long)
        private external fun reset(gifFileInPtr: Long): Boolean
        private external fun setSpeedFactor(gifFileInPtr: Long, factor: Float)
        private external fun getComment(gifFileInPtr: Long): String
        private external fun getLoopCount(gifFileInPtr: Long): Int
        private external fun setLoopCount(gifFileInPtr: Long, loopCount: Char)
        private external fun getSourceLength(gifFileInPtr: Long): Long
        private external fun getDuration(gifFileInPtr: Long): Int
        private external fun getCurrentPosition(gifFileInPtr: Long): Int
        private external fun seekToTime(gifFileInPtr: Long, position: Int, buffer: Bitmap)
        private external fun seekToFrame(gifFileInPtr: Long, frameNr: Int, buffer: Bitmap)
        private external fun saveRemainder(gifFileInPtr: Long)
        private external fun restoreRemainder(gifFileInPtr: Long): Long
        private external fun getAllocationByteCount(gifFileInPtr: Long): Long
        private external fun getMetadataByteCount(gifFileInPtr: Long): Long
        private external fun getNativeErrorCode(gifFileInPtr: Long): Int
        private external fun getCurrentFrameIndex(gifFileInPtr: Long): Int
        private external fun getCurrentLoop(gifFileInPtr: Long): Int
        private external fun postUnbindSurface(gifFileInPtr: Long)
        private external fun isAnimationCompleted(gifInfoPtr: Long): Boolean
        private external fun getSavedState(gifInfoPtr: Long): LongArray
        private external fun restoreSavedState(
            gifInfoPtr: Long,
            savedState: LongArray,
            mBuffer: Bitmap?
        ): Int

        private external fun getFrameDuration(gifInfoPtr: Long, index: Int): Int
        private external fun setOptions(gifInfoPtr: Long, sampleSize: Char, isOpaque: Boolean)
        private external fun getWidth(gifFileInPtr: Long): Int
        private external fun getHeight(gifFileInPtr: Long): Int
        private external fun getNumberOfFrames(gifInfoPtr: Long): Int
        private external fun isOpaque(gifInfoPtr: Long): Boolean
        private external fun startDecoderThread(gifInfoPtr: Long)
        private external fun stopDecoderThread(gifInfoPtr: Long)
        private external fun glTexImage2D(gifInfoPtr: Long, target: Int, level: Int)
        private external fun glTexSubImage2D(gifInfoPtr: Long, target: Int, level: Int)
        private external fun seekToFrameGL(gifInfoPtr: Long, index: Int)
        private external fun initTexImageDescriptor(gifInfoPtr: Long)
    }
}